boilerplate/app/components/AutoImage.tsx

```diff
@@ -1,26 +1,74 @@
 import React, { useLayoutEffect, useState } from "react"
 import { Image, ImageProps, ImageURISource, Platform } from "react-native"
 
+// MAVERICKTODO: document new props
+export interface AutoImageProps extends ImageProps {
+  /**
+   * How wide should the image be?
+   */
+  maxWidth?: number
+  /**
+   * How tall should the image be?
+   */
+  maxHeight?: number
+}
+
+/**
+ * A hook that will return the scaled dimensions of an image based on the
+ * provided dimesions' aspect ratio. If no desired dimensions are provided,
+ * it will return the original dimensions of the remote image.
+ *
+ * How is this different from `resizeMode: 'contain'`? Firstly, you can
+ * specify only one side's size (not both). Secondly, the image will scale to fit
+ * the desired dimensions instead of just being contained within its image-container.
+ *
+ * MAVERICKTODO: document hook
+ */
+
+export function useAutoImage(
+  remoteUri: string,
+  dimensions?: [maxWidth: number, maxHeight: number],
+): [width: number, height: number] {
+  const [[remoteWidth, remoteHeight], setRemoteImageDimensions] = useState([0, 0])
+  const remoteAspectRatio = remoteWidth / remoteHeight
+  const [maxWidth, maxHeight] = dimensions ?? []
+
+  useLayoutEffect(() => {
+    if (!remoteUri) return
+
+    Image.getSize(remoteUri, (w, h) => setRemoteImageDimensions([w, h]))
+  }, [remoteUri])
+
+  if (Number.isNaN(remoteAspectRatio)) return [0, 0]
+
+  if (maxWidth && maxHeight) {
+    const aspectRatio = Math.min(maxWidth / remoteWidth, maxHeight / remoteHeight)
+    return [remoteWidth * aspectRatio, remoteHeight * aspectRatio]
+  } else if (maxWidth) {
+    return [maxWidth, maxWidth / remoteAspectRatio]
+  } else if (maxHeight) {
+    return [maxHeight * remoteAspectRatio, maxHeight]
+  } else {
+    return [remoteWidth, remoteHeight]
+  }
+}
```

Comment:
This hook adds ability for automatic image scaling. This is integrated into AutoImage for remote image scaling, but can be updated to include the same functionality for local image as well (needs to call different method to get the size).